include/lin/references/tensor_stream_reference.hpp

```diff
@@ -0,0 +1,169 @@
+   *  errors will be triggered.
+   *
+   *  @sa internal::traits
+   */
+  constexpr TensorStreamReference(Stream<E> const &stream, size_t i, size_t j)
+  : _stream(stream), _i(i), _j(j) {
+    LIN_ASSERT((i >= 0) && (i + Traits::max_rows <= stream.rows()));
+    LIN_ASSERT((j >= 0) && (j + Traits::max_cols <= stream.cols()));
+
+    resize(Traits::max_rows, Traits::max_cols);
+  }
+
+  /** @brief Constructs a new reference with the provided mapping, anchor point,
+   *         and dimensions.
+   *
+   *  @param mapping Underlying stream.
+   *  @param i       Anchor point row index.
+   *  @param j       Anchor point column index.
+   *  @param r       Reference row dimension.
+   *  @param c       Reference column dimension.
+   *
+   *  The anchor point maps the top left corner of the reference to an element
+   *  in the underlying stream.
+   *
+   *  The reference traits must define a type that fits within the provided
+   *  mapping given the anchor point. If this is not the case, lin assertion
+   *  errors will be triggered.
+   *
+   *  Lin assertions errors will be triggered if the requested dimensions aren't
+   *  possible given the reference's traits.
+   *
+   *  @sa internal::traits
+   */
+  constexpr TensorStreamReference(Stream<E> const &stream, size_t i, size_t j, size_t r, size_t c)
+  : _stream(stream), _i(i), _j(j) {
+    LIN_ASSERT((i >= 0) && (i + r <= stream.rows()));
+    LIN_ASSERT((j >= 0) && (j + c <= stream.cols()));
+
+    resize(r, c);
+  }
+
+  /** @brief Provides read only access to tensor elements.
+   *
+   *  @param i Row index.
+   *  @param j Column index.
+   *
+   *  @return Value of the tensor element.
+   *
+   *  If the indices are out of bounds as defined by the reference's current
+   *  dimensions, lin assertion errors will be triggered.
+   *
+   *  If accessing data from a lazily evaluation tensor operation, you may want
+   *  to consider for the creation of a value backed type to reduce overhead.
+   *
+   *  @sa internal::Stream::eval
+   */
+  constexpr typename Traits::elem_t operator()(size_t i, size_t j) const {
+    LIN_ASSERT((i >= 0) && (i < rows()));
+    LIN_ASSERT((j >= 0) && (j < cols()));
+
+    return _stream(_i + i, _j + j);
+  }
+
+  /** @brief Provides read only access to tensor elements.
+   *
+   *  @param i Index.
+   *
+   *  @return Reference to the tensor element.
+   *
+   *  Element access proceeds as if all the elements of the tensor stream were
+   *  flattened into an array in row major order.
+   *
+   *  If the index is out of bounds as defined by the tensor's current size, lin
+   *  assertion errors will be triggered.
+   *
+   *  If accessing data from a lazily evaluation tensor operation, you may want
+   *  to consider for the creation of a value backed type to reduce overhead.
+   *
+   *  @sa internal::Stream::eval
+   */
+  constexpr typename Traits::elem_t operator()(size_t i) const {
```

Comment:
just curious, why did you choose operator() instead of []?

Reply:
The `operator[]` is implemented using the `operator()` in parent classes so child classes only need to provide `operator()`. Similarly, stream children provide only the `const` version while mapping children provide only the non-`const` version. `template <class D> lin::internal::Mapping` implements the `const` version with the non-`const` version.